### PR TITLE
Add parquet bloom filter read support

### PR DESCRIPTION
**Pull request by nmahadevuni**
This PR adds support to read a column's bloom filter if it exists and tests it to discard rowgroups that don't satisfy the filter. For e.g., if the query has a filter `c1 = 100` and a bloom filter for column c1 exists, the test is done on bloom filter whether value 100 exists in the bloom filter. If it returns true, then Parquet reader proceeds to read the column chunks for the rowgroup, else it discards that rowgroup.

This feature is supported using a configurable option `hive.parquet.use-bloom-filter`.

### CODE REVIEW

File: velox/type/AbstractBloomFilter.h [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+#pragma once
+
+#include <cstdint>
+#include <string>
+
+class AbstractBloomFilter {
+ public:
+  virtual bool mightContain(int32_t value) = 0;
+  virtual bool mightContain(int64_t value) = 0;
+  virtual bool mightContain(std::string value) = 0;
```
Comment (Yuhta) [inline velox/type/AbstractBloomFilter.h:26]: `const std::string&`

File: velox/type/Filter.h [hunk 2]
```diff
@@ -80,6 +81,16 @@ class Filter : public velox::ISerializable {
   // runtime.
   static constexpr bool deterministic = true;
 
+  virtual bool isSingleValue() const {
+    return false;
+  }
+
+  virtual bool testBloomFilter(
+      std::unique_ptr<AbstractBloomFilter>& bloomFilter,
```
Comment (Yuhta) [inline velox/type/Filter.h:89]: `const AbstractBloomFilter&`

File: velox/type/Filter.h [hunk 3]
```diff
@@ -1878,6 +1895,20 @@ class BytesValues final : public Filter {
...
+  }
+
+  bool isSingleValue() const override {
+    return isSingleValue_;
+  }
+
+  bool testBloomFilter(
+      std::unique_ptr<AbstractBloomFilter>& bloomFilter,
+      const velox::Type& type) const override {
+    return bloomFilter->mightContain(lower_);
```
Comment (Yuhta) [inline velox/type/Filter.h:1909]: I don't think this is correct, you need to loop over all `values_` here

File: velox/type/Filter.h [hunk 4]
```diff
@@ -80,6 +81,16 @@ class Filter : public velox::ISerializable {
   // runtime.
   static constexpr bool deterministic = true;
 
+  virtual bool isSingleValue() const {
```
Comment (Yuhta) [inline velox/type/Filter.h:84]: What is this used for?

File: velox/type/Filter.h (hunk 3 above)
Comment (nmahadevuni) [reply velox/type/Filter.h:1909]: We can apply bloom filter only for exact lookup(equality predicate) and not range of values, in which case `lower_` and `upper_` will be the same. `isSingleValue` method is used for determining this. The check for single value is made in the caller method `rowGroupMatches()`

Comment (Yuhta) [reply velox/type/Filter.h:1909]: If it is only supposed to be called with single value filter, please document it in the interface method and check the filter is single valued inside the method (remove `isSingleValue()`), instead of exposing this implementation details to file readers:

```c++
bool testBloomFilter(
      std::unique_ptr<AbstractBloomFilter>& bloomFilter,
      const velox::Type& type) const override {
  return lower_ < upper_ || bloomFilter->mightContain(lower_);
}
```

Comment (nmahadevuni) [reply velox/type/Filter.h:1909]: `isSingleValue()` was already implemented as a public method before my changes for some Filter's sub classes. I just added an interface method in Filter class added overrides since this can be used for this PR's purpose.

Comment (nmahadevuni) [reply velox/type/Filter.h:1909]: I will move the call to isSingleValue() inside the testBloomFilter().

**Review by yingsu00 — COMMENTED** [review]
Let's first try to remove the Arrow dependencies.

**Comment by nmahadevuni** [comment]
@yingsu00 Can you please review this?

**Review by yingsu00 — COMMENTED** [review]
@nmahadevuni This only made bigint range filters work. 

- What about in lists? These are cases where bloom filter is most useful. 
- What about other types? e.g. Hugeint, bool, varchar?
- What about multi ranges?

File: velox/type/AbstractBloomFilter.h [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+ */
+
+#pragma once
+
+#include <cstdint>
+#include <string>
+
+class AbstractBloomFilter {
+ public:
+  virtual bool mightContain(int32_t value) = 0;
```
Comment (Yuhta) [inline velox/type/AbstractBloomFilter.h:24]: Why not making these methods `const`?  As far as I see those `FindHash` is already `const`

File: velox/type/Filter.h [hunk 6]
```diff
@@ -80,6 +81,12 @@ class Filter : public velox::ISerializable {
   // runtime.
   static constexpr bool deterministic = true;
 
+  virtual bool testBloomFilter(
+      AbstractBloomFilter& bloomFilter,
```
Comment (Yuhta) [inline velox/type/Filter.h:85]: Than you can take `const AbstractBloomFilter&` here

File: velox/dwio/parquet/reader/ParquetReader.cpp [hunk 7]
```diff
@@ -629,7 +635,17 @@ ParquetRowReader::ParquetRowReader(
   if (rowGroups_.empty()) {
     return; // TODO
   }
-  ParquetParams params(pool_, columnReaderStats_, readerBase_->fileMetaData());
+
+  if (readerBase_->isUseParquetBloomFilter()) {
+    bloomFilterReader_ = BloomFilterReader::make(
```
Comment (yingsu00) [inline velox/dwio/parquet/reader/ParquetReader.cpp:640]: If there is no bloom filter in any row group or column chunk, there is no need to create a bloomFilterReader_. 
Actually, I wonder why Arrow used so many classes BloomFilterReader, BloomFilterReaderImpl, RowGroupBloomFilterReader, RowGroupBloomFilterReaderImpl and BloomFilter. It adds several levels to trace when people reads the source code and not quite strait-forward. The only important and necessary structure is BloomFilter. Can we just have a simple function in ParquetData
```
std::unique_ptr<BloomFilter> getBloomFilter(const thrift::RowGroup& rowgroup, uint32_t column)
```
that just read the bloom filter for the specified rowgroup and column? Then we can remove bloomFilterReader_ and only lazily read the BloomFilter in `ParquetData::rowGroupMatches(uint32_t rowGroupId, common::Filter* FOLLY_NULLABLE filter)` and no need to pass in the rowGroupBloomFilterReader

File: velox/type/AbstractBloomFilter.h [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cstdint>
+#include <string>
+
+class AbstractBloomFilter {
```
Comment (yingsu00) [inline velox/type/AbstractBloomFilter.h:22]: Why introduce this class?

File: velox/dwio/parquet/reader/ParquetBloomFilter.h [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+ */
+
+#pragma once
+
+#include "velox/dwio/parquet/reader/BloomFilter.h"
+#include "velox/type/AbstractBloomFilter.h"
+
+namespace facebook::velox::parquet {
+
+class ParquetBloomFilter final : public AbstractBloomFilter {
```
Comment (yingsu00) [inline velox/dwio/parquet/reader/ParquetBloomFilter.h:24]: Any other Velox XXXBloomFilter that will extend AbstractBloomFilter? It's good to have the AbstractBloomFilter interface, but is it necessary to have a separate ParquetBloomFilter? The original Arrow BloomFilter class can have other sub classes, but I don't see if other Velox XXXBloomFilter would have different implementation than ParquetBloomFilter.

File: velox/dwio/parquet/reader/ParquetData.cpp [hunk 10]
```diff
@@ -73,15 +83,36 @@ bool ParquetData::rowGroupMatches(
     return true;
   }
 
+  bool statsMatched = true;
+  bool filterRangeCoversStatsRange = false;
```
[the remark this answers is not in the document]
Comment (yingsu00) [reply velox/dwio/parquet/reader/ParquetData.cpp:87]: Actually I think it's better not to have these variables. With the previous suggestions, this function an be simplified to
```
bool ParquetData::rowGroupMatches(
    uint32_t rowGroupId,
    common::Filter* FOLLY_NULLABLE filter) {
  auto column = type_->column();
  auto type = type_->type();
  auto rowGroup = rowGroups_[rowGroupId];
  assert(!rowGroup.columns.empty());

  if (!filter) {
    return true;
  }

  if (rowGroup.columns[column].__isset.meta_data &&
      rowGroup.columns[column].meta_data.__isset.statistics) {
    auto columnStats = buildColumnStatisticsFromThrift(
        rowGroup.columns[column].meta_data.statistics,
        *type,
        rowGroup.num_rows);

    if (!testFilter(filter, columnStats.get(), rowGroup.num_rows, type)) {
      return false;
    }

    if (isUseBloomFilter() &&
        isFilterRangeCoversStatsRange(filter, columnStats.get(), type) &&
        rowGroup.columns[column].__isset.meta_data &&
        rowGroup.columns[column].meta_data.__isset.bloom_filter_offset) {
      std::unique_ptr<AbstractBloomFilter> bloomFilter =
          getColumnBloomFilter(rowGroup, column);
      return filter->testBloomFilter(*bloomFilter, *type);
    }
  }
}
```

File: velox/dwio/parquet/reader/ParquetBloomFilter.h (hunk 9 above)
Comment (Yuhta) [reply velox/dwio/parquet/reader/ParquetBloomFilter.h:24]: It's more for dependency management.  We cannot have parquet dependency in `velox/type`

File: velox/dwio/parquet/reader/BloomFilterReader.cpp [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+ * limitations under the License.
+ */
+
+// Adapted from Apache Arrow.
+
+#include "velox/dwio/parquet/reader/BloomFilterReader.h"
+
+namespace facebook::velox::parquet {
+
+class RowGroupBloomFilterReaderImpl final : public RowGroupBloomFilterReader {
```
[the remark this answers is not in the document]
Comment (nmahadevuni) [reply velox/dwio/parquet/reader/BloomFilterReader.cpp:23]: This is a big refactor. The ones in the writer/arrow/tests use the arrow implementation. I will do it in a separate PR after we merge this one.

File: velox/dwio/parquet/reader/ParquetBloomFilter.h (hunk 9 above)
Comment (nmahadevuni) [reply velox/dwio/parquet/reader/ParquetBloomFilter.h:24]: From previous discussion on https://github.com/facebookincubator/velox/issues/4414, we expect to use the same interface for ORC/DWRF bloom filter implementation.

File: velox/dwio/parquet/reader/ParquetData.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (nmahadevuni) [reply velox/dwio/parquet/reader/ParquetData.cpp:87]: I had to modify it a little since we may not have column stats present but bloom filter present, so I have two separate if blocks, not embedded.

File: velox/dwio/common/ScanSpec.cpp [hunk 12]
```diff
@@ -369,6 +369,47 @@ bool testFilter(
   return true;
 }
 
+bool isFilterRangeCoversStatsRange(
+    common::Filter* filter,
+    dwio::common::ColumnStatistics* stats,
+    const TypePtr& type) {
+  switch (type->kind()) {
```
[the remark this answers is not in the document]
Comment (nmahadevuni) [reply velox/dwio/common/ScanSpec.cpp:376]: I'm moving it to ParquetData.cpp and added a comment at the call site. Let us keep this name since its closer to the actual logic.

File: velox/type/AbstractBloomFilter.h (hunk 8 above)
Comment (nmahadevuni) [reply velox/type/AbstractBloomFilter.h:22]: Please see below for comment.

File: velox/dwio/parquet/reader/ParquetReader.cpp (hunk 7 above)
Comment (nmahadevuni) [reply velox/dwio/parquet/reader/ParquetReader.cpp:640]: We need file input stream to create the bloom filter reader which is not available in ParquetData. Will create the BloomFilterReader, call the function to get the RowGroupBloomFilterReader only when required.

File: velox/dwio/parquet/reader/ParquetData.cpp [hunk 13]
```diff
@@ -143,4 +212,42 @@ std::pair<int64_t, int64_t> ParquetData::getRowGroupRegion(
...
+  auto fileSize = bufferedInput.getInputStream()->getLength();
+  if (fileSize <= bloomFilterOffset) {
+    VELOX_FAIL("file size less or equal than bloom offset");
+  }
+
+  auto inputStream = bufferedInput.read(
+      bloomFilterOffset,
+      fileSize - bloomFilterOffset,
+      dwio::common::LogType::FOOTER);
+  auto bloomFilter =
```
Comment (yingsu00) [inline velox/dwio/parquet/reader/ParquetData.cpp:247]: Every time this function is called, it would read and deserialize the bloom filter again. I think it'd be better to make the bloom filter as a class member, and if it's read before, just return it.

File: velox/dwio/parquet/reader/ParquetReader.cpp [hunk 14]
```diff
@@ -748,6 +764,17 @@ bool ParquetRowReader::isRowGroupBuffered(int32_t rowGroupIndex) const {
   return readerBase_->isRowGroupBuffered(rowGroupIndex);
 }
 
+std::unique_ptr<BloomFilter> ParquetRowReader::getBloomFilter(
+    const uint32_t rowGroupId,
+    const uint32_t column) {
+  return facebook::velox::parquet::getBloomFilter(
```
Comment (yingsu00) [inline velox/dwio/parquet/reader/ParquetReader.cpp:770]: This looks a bit hacky to me. It will read and deserialize the bloom filter every time it's called for the same rowGroupId and column, though the callers are not doing this so far. When a user sees an interface like this, he or she may naturally think the  ParquetData::getBloomFilter() would not be doing the work and should have a single entry point, which is when you filter the row groups. But the callsite here is reading and deserializing the bloom filter again. I understand this is just for tests, but the existence of this function here would not suggest that, but instead, it looks like some real business logic code. If we make the get the bloom filter as a member of ParquetData, here we can just get the bloom filter by getting its the column'th child SelectiveColumnReader and the ParquetData it owns, then we can get the Bloom filter.

File: velox/dwio/parquet/reader/ParquetData.cpp [hunk 15]
```diff
@@ -143,4 +212,42 @@ std::pair<int64_t, int64_t> ParquetData::getRowGroupRegion(
...
+  if (column < 0 || column >= rowGroupMetadata.columns.size()) {
+    VELOX_FAIL("Invalid column index at column ordinal ", column);
+  }
+
+  auto colChunk = rowGroupMetadata.columns[column];
+  if (colChunk.__isset.crypto_metadata) {
+    VELOX_NYI("Cannot read encrypted bloom filter yet");
+  }
+
+  if (!colChunk.meta_data.__isset.bloom_filter_offset) {
```
[the remark this answers is not in the document]
Comment (nmahadevuni) [reply velox/dwio/parquet/reader/ParquetData.cpp:234]: I would like to keep it here so for any tests that we write, we get a null bloom filter rather than trying to read a non-existent bloom filter. Will change the call site.

[the remark this answers is not in the document]
Comment (nmahadevuni) [reply velox/dwio/parquet/reader/ParquetData.cpp:234]: Will retain this check at the call site too, so we can assume that bloom filter should always be present if the offset is set.

File: velox/dwio/parquet/reader/ParquetData.cpp [hunk 16]
```diff
@@ -143,4 +210,59 @@ std::pair<int64_t, int64_t> ParquetData::getRowGroupRegion(
   return {fileOffset, length};
 }
 
+std::shared_ptr<BloomFilter> ParquetData::getBloomFilter(
+    dwio::common::BufferedInput& bufferedInput,
+    const uint32_t rowGroupId,
+    const uint32_t column) {
```
Comment (yingsu00) [inline velox/dwio/parquet/reader/ParquetData.cpp:216]: No need to pass column parameter. The column information is already in this object.

File: velox/dwio/parquet/common/BloomFilter.cpp [hunk 17]
```diff
@@ -0,0 +1,239 @@
...
+  }
+}
+
+BlockSplitBloomFilter BlockSplitBloomFilter::deserialize(
+    dwio::common::SeekableInputStream* input,
+    memory::MemoryPool& pool) {
+  const void* headerBuffer;
+  int32_t size;
+  input->Next(&headerBuffer, &size);
+  const char* bufferStart = reinterpret_cast<const char*>(headerBuffer);
```
[the remark this answers is not in the document]
Comment (nmahadevuni) [reply velox/dwio/parquet/common/BloomFilter.cpp:107]: Cannot inline these since `ThriftStreamingTransport` constructor takes references for these two arguments. `const char*& bufferStart`

File: velox/dwio/parquet/common/BloomFilter.cpp [hunk 18]
```diff
@@ -0,0 +1,239 @@
...
+  const auto bloomFilterBytesInHeader = size - headerSize;
+  if (bloomFilterBytesInHeader > 0) {
+    std::memcpy(
+        buffer->asMutable<char>(),
+        reinterpret_cast<const uint8_t*>(headerBuffer) + headerSize,
+        bloomFilterBytesInHeader);
+  }
+  const auto requiredReadSize = bloomFilterSize - bloomFilterBytesInHeader;
+
+  input->readFully(buffer->asMutable<char>(), requiredReadSize);
```
[the remark this answers is not in the document]
Comment (nmahadevuni) [reply velox/dwio/parquet/common/BloomFilter.cpp:142]: This was a bug, fixed it and added a test case.

**Comment by nmahadevuni** [comment]
@yingsu00 Thanks, I have addressed your comments. Please review!

File: velox/dwio/parquet/reader/ParquetData.h [hunk 19]
```diff
@@ -199,6 +217,13 @@ class ParquetData : public dwio::common::FormatData {
   // ahead of first use, not at construction.
   std::vector<std::unique_ptr<dwio::common::SeekableInputStream>> streams_;
 
+  bool isParquetBloomFilterEnabled_;
+  // RowGroup+Column to BloomFilter map
+  std::unordered_map<
+      uint32_t,
+      std::unordered_map<uint32_t, std::shared_ptr<BloomFilter>>>
```
Comment (yingsu00) [inline velox/dwio/parquet/reader/ParquetData.h:224]: In my understanding, we don't need to use a 2-level map here. Each column reader own's its ParquetData object by itself, therefore it doesn't need to store the column index. Would you please try just store <RowGroupId, BloomFilter>?

**Comment by nmahadevuni** [comment]
@yingsu00 I have addressed the comments. Please review.

**Comment by pedroerp** [comment]
Is this PR still a draft? If not, could we update the title?

File: velox/dwio/parquet/common/BloomFilter.cpp [hunk 20]
```diff
@@ -0,0 +1,247 @@
...
+
+#include <cstdint>
+#include <cstring>
+#include <memory>
+
+namespace facebook::velox::parquet {
+
+constexpr uint32_t BlockSplitBloomFilter::SALT[kBitsSetPerBlock];
+
+BlockSplitBloomFilter::BlockSplitBloomFilter(memory::MemoryPool* pool)
```
Comment (yingsu00) [inline velox/dwio/parquet/common/BloomFilter.cpp:34]: The BloomFilter class was moved from writer/arrow, but velox/dwio/parquet/writer/arrow/tests/BloomFilter.cpp is still there. Will you be able to send a separate commit/PR to remove the duplicate code there?

Comment (nmahadevuni) [reply velox/dwio/parquet/common/BloomFilter.cpp:34]: Yes. I will remove the ones in writer/arrow once this PR is merged.

**Comment by nmahadevuni** [comment]
@pedroerp @Yuhta Can you please review this PR ?

File: velox/common/base/AbstractBloomFilter.h [hunk 21]
```diff
@@ -0,0 +1,28 @@
...
+ */
+
+#pragma once
+
+#include <cstdint>
+#include <string>
+
+class AbstractBloomFilter {
+ public:
+  virtual bool mightContain(int32_t value) const = 0;
```
Comment (pedroerp) [inline velox/common/base/AbstractBloomFilter.h:24]: does that mean that checking existence will trigger a virtual call? Won't this be terribly slow?

File: velox/connectors/hive/HiveConfig.h [hunk 22]
```diff
@@ -104,6 +104,9 @@ class HiveConfig {
   static constexpr const char* kOrcUseColumnNamesSession =
       "hive_orc_use_column_names";
 
+  static constexpr const char* kEnableParquetBloomFilter =
```
Comment (pedroerp) [inline velox/connectors/hive/HiveConfig.h:107]: what does this option do? Does it generate bloom filter when writing files, or uses them when reading files with bloom filter? Regardless, we should add comments and a PR description to clarify.

File: velox/connectors/hive/HiveConnector.cpp [hunk 23]
```diff
@@ -82,6 +82,21 @@ std::unique_ptr<DataSource> HiveConnector::createDataSource(
         std::string,
         std::shared_ptr<connector::ColumnHandle>>& columnHandles,
     ConnectorQueryCtx* connectorQueryCtx) {
+  dwio::common::ReaderOptions options(connectorQueryCtx->memoryPool());
```
Comment (pedroerp) [inline velox/connectors/hive/HiveConnector.cpp:85]: I'm probably missing something, but how/where do we use this `options` object?

File: velox/dwio/common/CMakeLists.txt [hunk 24]
```diff
@@ -77,4 +77,5 @@ target_link_libraries(
   velox_exec
   Boost::regex
   Folly::folly
-  glog::glog)
+  glog::glog
+  ${Protobuf_LIBRARIES})
```
Comment (pedroerp) [inline velox/dwio/common/CMakeLists.txt:81]: protobuf is parquet dependency, so we shouldn't have it in the common library.

File: velox/dwio/common/Options.h [hunk 25]
```diff
@@ -389,6 +389,7 @@ class ReaderOptions : public io::ReaderOptions {
   bool fileColumnNamesReadAsLowerCase{false};
   bool useColumnNamesForColumnMapping_{false};
   std::shared_ptr<folly::Executor> ioExecutor_;
+  bool enableParquetBloomFilter_{false};
```
Comment (pedroerp) [inline velox/dwio/common/Options.h:392]: I saw @yingsu00 had a comment about this; can we have the bloom filter stuff being a Parquet option, and not a general dwio option (since this is specific to Parquet)?

File: velox/dwio/parquet/common/BloomFilter.cpp (hunk 20 above)
Comment (pedroerp) [reply velox/dwio/parquet/common/BloomFilter.cpp:34]: Can we do this move/cleanup in a different PR? It makes this one unnecessary large and harder to review.

**Review by pedroerp — CHANGES_REQUESTED** [review]
Made some initial comments, but let's please first move the Arrow stuff needed in a separate PR to make this one smaller. Let's also add a PR summary explaining what exact this PR is adding, the options, the behavior, etc.

**Comment by nmahadevuni** [comment]
Sorry for the delay. I created another PR https://github.com/facebookincubator/velox/pull/10516 with the arrow code moved to velox, as per the reviewers comment here. Once that is merged, I will restart work on this PR. @yingsu00 @Yuhta @pedroerp

**Comment by nmahadevuni** [comment]
reopening

**Comment by nmahadevuni** [comment]
Rebased on top of the first PR with arrow changes.

File: velox/common/base/AbstractBloomFilter.h (hunk 21 above)
Comment (nmahadevuni) [reply velox/common/base/AbstractBloomFilter.h:24]: This is done once per row group, so it should be ok? Also, we are creating method 
```
virtual bool testBloomFilter(
      const AbstractBloomFilter& bloomFilter,
      const velox::Type& type) 
```
in the `Filter` class so, we need to have an abstract type defined.

File: velox/dwio/common/Options.h (hunk 25 above)
Comment (nmahadevuni) [reply velox/dwio/common/Options.h:392]: I renamed this to generic `readBloomFilter_`, so it can be used for any format. Also, I'm currently initializing this only for parquet in `configureReaderOptions`.
